### PR TITLE
add instantiate2 broadcast in SigningCosmwasmClient

### LINKED ISSUE #1399: Add SigningCosmWasmClient.instantiate2
Right now, CosmJS can only pre-calculate instantiate2 addresses but not do the actual instantiation.

### PR DESCRIPTION
**Pull request by loin3**
Closes #1399

### CODE REVIEW

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts [hunk 1]
```diff
@@ -103,6 +105,38 @@ export interface InstantiateOptions {
...
+  /**
+   * salt is an arbitrary value provided by the sender. Size can be 1 to 64.
+   */
+  readonly salt?: Uint8Array;
+
+  /**
+   * FixMsg include the msg value into the hash for the predictable address.
+   * Default is false
+   */
+  readonly fixMsg?: boolean;
```
Comment (webmaster128) [inline packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:137]: This should be removed (and hardcoded to false)

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts [hunk 2]
```diff
@@ -335,6 +369,44 @@ export class SigningCosmWasmClient extends CosmWasmClient {
...
+      typeUrl: "/cosmwasm.wasm.v1.MsgInstantiateContract2",
+      value: MsgInstantiateContract2.fromPartial({
+        sender: senderAddress,
+        codeId: Long.fromString(new Uint53(codeId).toString()),
+        label: label,
+        msg: toUtf8(JSON.stringify(msg)),
+        funds: [...(options.funds || [])],
+        admin: options.admin,
+        salt: options.salt,
+        fixMsg: options.fixMsg,
```
Comment (webmaster128) [inline packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:390]:
```suggestion
        fixMsg: false,
```

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts [hunk 3]
```diff
@@ -103,6 +105,38 @@ export interface InstantiateOptions {
...
+  /**
+   * A bech32 encoded address of an admin account.
+   * Caution: an admin has the privilege to upgrade a contract. If this is not desired, do not set this value.
+   */
+  readonly admin?: string;
+
+  /**
+   * salt is an arbitrary value provided by the sender. Size can be 1 to 64.
+   */
+  readonly salt?: Uint8Array;
```
Comment (webmaster128) [inline packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:131]: This value is required. Let's add it as a required positional argument after `    codeId: number,`

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts [hunk 4]
```diff
@@ -335,6 +369,44 @@ export class SigningCosmWasmClient extends CosmWasmClient {
     };
   }
 
+  public async instantiate2(
+    senderAddress: string,
+    codeId: number,
+    msg: Record<string, unknown>,
```
Comment (webmaster128) [inline packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:375]:
```suggestion
    msg: JsonObject,
```

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts [hunk 5]
```diff
@@ -103,6 +105,38 @@ export interface InstantiateOptions {
   readonly admin?: string;
 }
 
+/**
+ * The options of an .instantiate() call.
+ * All properties are optional.
+ */
+export interface Instantiate2Options {
```
Comment (webmaster128) [inline packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:112]: This should be listed in `index.ts`

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts [hunk 6]
```diff
@@ -103,6 +105,38 @@ export interface InstantiateOptions {
   readonly admin?: string;
 }
 
+/**
+ * The options of an .instantiate() call.
```
Comment (webmaster128) [inline packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:109]:
```suggestion
 * The options of an .instantiate2() call.
```

File: packages/cosmwasm-stargate/src/modules/wasm/messages.ts [hunk 7]
```diff
@@ -39,6 +41,19 @@ export function isMsgInstantiateContractEncodeObject(
   );
 }
 
+export interface MsgInstantiateContract2EncodeObject extends EncodeObject {
+  readonly typeUrl: "/cosmwasm.wasm.v1.MsgInstantiateContract2";
+  readonly value: Partial<MsgInstantiateContract2>;
+}
+
+export function isMsgInstantiateContract2EncodeObject(
```
Comment (webmaster128) [inline packages/cosmwasm-stargate/src/modules/wasm/messages.ts:49]: Those two symbols should also be listed in `index.ts`

**Review by webmaster128 — COMMENTED** [review]
Very nice, thank you!

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts (hunk 5 above)
Comment (loin3) [reply packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:112]: I found that `Instantiate2Options` is exactly same with `InstantiateOptions` after I delete salt and fixMsg.
So I delete `Instantiate2Options`.

File: packages/cosmwasm-stargate/src/signingcosmwasmclient.ts (hunk 6 above)
Comment (loin3) [reply packages/cosmwasm-stargate/src/signingcosmwasmclient.ts:109]: I found that `Instantiate2Options` is exactly same with `InstantiateOptions` after I delete salt and fixMsg.
So I delete `Instantiate2Options`.

**Comment by webmaster128** [comment]
Thank you. Will be shipped in the next release.

**Pull request merged**